examples/vision/barlow_twins.py

```diff
@@ -0,0 +1,990 @@
+    """
+
+    def identity_block(self, x, filter):
+        # copy tensor to variable called x_skip
+        x_skip = x
+        # Layer 1
+        x = tf.keras.layers.Conv2D(filter, (3, 3), padding="same")(x)
+        x = tf.keras.layers.BatchNormalization(axis=3)(x)
+        x = tf.keras.layers.Activation("relu")(x)
+        # Layer 2
+        x = tf.keras.layers.Conv2D(filter, (3, 3), padding="same")(x)
+        x = tf.keras.layers.BatchNormalization(axis=3)(x)
+        # Add Residue
+        x = tf.keras.layers.Add()([x, x_skip])
+        x = tf.keras.layers.Activation("relu")(x)
+        return x
+
+    def convolutional_block(self, x, filter):
+        # copy tensor to variable called x_skip
+        x_skip = x
+        # Layer 1
+        x = tf.keras.layers.Conv2D(filter, (3, 3), padding="same", strides=(2, 2))(x)
+        x = tf.keras.layers.BatchNormalization(axis=3)(x)
+        x = tf.keras.layers.Activation("relu")(x)
+        # Layer 2
+        x = tf.keras.layers.Conv2D(filter, (3, 3), padding="same")(x)
+        x = tf.keras.layers.BatchNormalization(axis=3)(x)
+        # Processing Residue with conv(1,1)
+        x_skip = tf.keras.layers.Conv2D(filter, (1, 1), strides=(2, 2))(x_skip)
+        # Add Residue
+        x = tf.keras.layers.Add()([x, x_skip])
+        x = tf.keras.layers.Activation("relu")(x)
+        return x
+
+    def __call__(self, shape=(32, 32, 3)):
+        # Step 1 (Setup Input Layer)
+        x_input = tf.keras.layers.Input(shape)
+        x = tf.keras.layers.ZeroPadding2D((3, 3))(x_input)
+        # Step 2 (Initial Conv layer along with maxPool)
+        x = tf.keras.layers.Conv2D(64, kernel_size=7, strides=2, padding="same")(x)
+        x = tf.keras.layers.BatchNormalization()(x)
+        x = tf.keras.layers.Activation("relu")(x)
+        x = tf.keras.layers.MaxPool2D(pool_size=3, strides=2, padding="same")(x)
+        # Define size of sub-blocks and initial filter size
+        block_layers = [3, 4, 6, 3]
+        filter_size = 64
+        # Step 3 Add the Resnet Blocks
+        for i in range(4):
+            if i == 0:
+                # For sub-block 1 Residual/Convolutional block not needed
+                for j in range(block_layers[i]):
+                    x = self.identity_block(x, filter_size)
+            else:
+                # One Residual/Convolutional Block followed by Identity blocks
+                # The filter size will go on increasing by a factor of 2
+                filter_size = filter_size * 2
+                x = self.convolutional_block(x, filter_size)
+                for j in range(block_layers[i] - 1):
+                    x = self.identity_block(x, filter_size)
+        # Step 4 End Dense Network
+        x = tf.keras.layers.AveragePooling2D((2, 2), padding="same")(x)
+        x = tf.keras.layers.Flatten()(x)
+        model = tf.keras.models.Model(inputs=x_input, outputs=x, name="ResNet34")
+        return model
+
+
+def build_twin() -> tf.keras.Model:
```

Comment:
The model-class (`ResNet34`) and `build_twin` function can be done at a time. It'll optimize the code.

Reply:
I will try this out. One concern is that it combines the resnet architecture with the main model(which makes it a little overwhelming to be honest). I thought it would be good if i could separate them because of replacement(it is easy to replace resnet34 with resnet50 or resnet20 if they are separated). What do you think?

Reply:
Youo may also consider building the model within the `BarlowModel` class where we can import `ResNet50` from `keras.applications`.  Something like [this](https://colab.research.google.com/drive/1Wc56C-_4nftQV6YMDj-SdlbRgUnFKvin?usp=sharing).

Reply:
Hello, thanks for the reply. That was my original solution, though I found that a resnet34 gives significant boosts in accuracy(59% to 64%) and also considerable decrease in training time(2 hours to 1 hr 30 minutes)

Reply:
Let me know though if there is a keras implementation of Resnet34

Reply:
Yes, there are actually, [here](https://github.com/qubvel/classification_models#architectures), unofficial but trusted. ([#15269](https://github.com/keras-team/keras/issues/15269))

Reply:
hmm, i see. looks like keras-contrib is deprecated, and tensorflow-addons is recommended(and i actually use it). Is your model implementation there?